### PR TITLE
Add tracing extra fields to logging callback

### PR DESCRIPTION
**Pull request by packgron**
### Problem
Until now some of information was lost then telio was used from ffi api, because non of the extra fields there send via telio log callback.

### Solution
Switched to tracing subscriber based implementation, as it correctly implements span logic, and will allow us to add additional systems in the future using layer approach.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

File: src/ffi/logging/test.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use crate::{TelioLogLevel, TelioLoggerCb};
+
+#[test]
+fn test_trace_via_telio_cb() {
+    let expected = [(
```
Comment (mathiaspeters) [inline src/ffi/logging/test.rs:15]: It doesn't seem like this is used

Comment (packgron) [reply src/ffi/logging/test.rs:15]: Ah missed assert somehow

Comment (packgron) [reply src/ffi/logging/test.rs:15]: Forgot to commit

Comment (packgron) [reply src/ffi/logging/test.rs]: yeh forgot to remove, though originaly that it would be simpler to maintain, but also so  did not see tests/loger.rs,
but i prefer to have tests next to code

File: tests/logger.rs [hunk 2]
```diff
@@ -26,7 +25,7 @@ mod test_module {
                 payload: String,
             ) -> FFIResult<()> {
                 assert!(matches!(log_level, telio::ffi_types::TelioLogLevel::Info));
-                assert_eq!(r#""logger::test_module":43 test message"#, payload);
+                assert_eq!(r#""logger::test_module":42 test message"#, payload);
```
Comment (matislovas) [inline tests/logger.rs:28]: Where does this number comes from? Maybe some named const for it ..?

**Comment by matislovas** [comment]
+1

File: src/ffi/logging.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    if log_status.counter < 10 {
+        log_status.counter += 1;
+        return Some(msg);
+    }
+
+    log_status.counter += 1;
+    None
+}
```
Comment (stalowyjez) [inline src/ffi/logging.rs:161]: It seems that some code is just moved (like this function) - I think it makes things a lot simpler, when such moving is kept in separate commits.

File: src/ffi/logging.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            logs.push((level, payload));
+            Ok(())
+        }
+    }
+
+    impl Debug for Log {
+        fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            Ok(())
+        }
+    }
```
Comment (stalowyjez) [inline src/ffi/logging.rs:235]: As I understand you need it only for TelioLoggerCb trait?

File: src/ffi/logging.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    #[derive(Default, Clone)]
+    struct Log(Arc<Mutex<Vec<(TelioLogLevel, String)>>>);
+    impl TelioLoggerCb for Log {
+        fn log(&self, level: TelioLogLevel, payload: String) -> crate::FFIResult<()> {
+            let mut logs = self.0.lock().expect("ok");
+            logs.push((level, payload));
+            Ok(())
+        }
+    }
```
Comment (stalowyjez) [inline src/ffi/logging.rs:229]: Nit: I think we usually keep such stuff before the tests, maybe it would be nice to follow that

File: src/ffi/logging.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            meta.line().unwrap_or(0),
+        )?;
+
+        ctx.format_fields(writer.by_ref(), event)?;
+
+        writeln!(writer)
+    }
+}
+
+pub struct FfiCallback {
```
Comment (stalowyjez) [inline src/ffi/logging.rs:66]: We have `FFIResult`, so I guess it would be nice to keep it consistent and name it `FFICallback`

File: src/ffi/logging.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            callback: logger.into(),
+        }
+    }
+}
+
+impl<'a> MakeWriter<'a> for FfiCallback {
+    type Writer = FfiCallbackWriter;
+
+    fn make_writer(&self) -> Self::Writer {
+        unreachable!("make_writer_for is used")
```
Comment (stalowyjez) [inline src/ffi/logging.rs:81]: This message sounds weird in this context imho,  maybe something like "make_writer is not expected to be used"?

File: tests/logger.rs (hunk 2 above)
Comment (packgron) [reply tests/logger.rs:28]: So this is the line of info!() trace at the top, for my test I made it relative, i will make it smarter here also

File: src/ffi/logging.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            // But this makes it more univesal for the future, then we dicide to
+            // switch to other formats
+            .map(|msg| msg.trim().to_string())
+            .unwrap_or_else(|err| format!("failed to decode trace as utf8 bytes: {err:?}"));
+
+        if let Some(filtered_msg) = filter_log_message(msg) {
+            let _ = self.cb.log(self.level, filtered_msg);
+        }
+
+        Ok(buf.len())
```
Comment (stalowyjez) [inline src/ffi/logging.rs:116]: Ehh, a nice amount of ugly boilerplate we have here

File: src/ffi/logging.rs (hunk 3 above)
Comment (packgron) [reply src/ffi/logging.rs:161]: Yes that's correct, will take note next time.

File: src/ffi/logging.rs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        // See https://github.com/tokio-rs/tracing/blob/tracing-subscriber-0.3.18/tracing-subscriber/src/fmt/fmt_layer.rs#L975
+        //
+        // Additional test added to double protect from future breakages
+
+        let msg = from_utf8(buf)
+            // Trim could be avoided by removing writeln! in formatter
+            // But this makes it more univesal for the future, then we dicide to
+            // switch to other formats
+            .map(|msg| msg.trim().to_string())
+            .unwrap_or_else(|err| format!("failed to decode trace as utf8 bytes: {err:?}"));
```
Comment (stalowyjez) [inline src/ffi/logging.rs:110]: If this function returns an error maybe it would be nice to return it on failed decoding?

File: src/ffi/logging.rs [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+        if let Some(filtered_msg) = filter_log_message(msg) {
+            let _ = self.cb.log(self.level, filtered_msg);
```
Comment (stalowyjez) [inline src/ffi/logging.rs:113]: And maybe here as well?

File: tests/logger.rs (hunk 2 above)
Comment (packgron) [reply tests/logger.rs:28]: Ok so turning this is no an easy task here, how currently tests are design, this will have to be left as is. Once cell would i think just add additional confusion

File: src/ffi/logging.rs (hunk 4 above)
Comment (packgron) [reply src/ffi/logging.rs:235]: Yes, but i see i can add a derive so will use it that

File: src/ffi/logging.rs (hunk 5 above)
Comment (packgron) [reply src/ffi/logging.rs:229]: I prefer then tests are front and center, and glue is out of sight preferably, will leave as is

File: src/ffi/logging.rs (hunk 6 above)
Comment (packgron) [reply src/ffi/logging.rs:66]: Well FFIResult is named againts rust naming standart :upside_down_face:

Comment (packgron) [reply src/ffi/logging.rs:66]: https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md#fine-points

Comment (stalowyjez) [reply src/ffi/logging.rs:66]: Hmm, so maybe we could change that name then?

File: src/ffi/logging.rs (hunk 8 above)
Comment (packgron) [reply src/ffi/logging.rs]: It's not that bad :D

File: src/ffi/logging.rs (hunk 7 above)
Comment (packgron) [reply src/ffi/logging.rs]: fixed

File: .unreleased/LLT-5430 [hunk 11]
```diff
@@ -0,0 +1 @@
+Aditional information fields are now shown in logs
```
[the remark this answers is not in the document]
Comment (packgron) [reply .unreleased/LLT-5430]: done

File: src/ffi/logging.rs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+                TelioLogLevel::Warning,
+                format!(
+                    "{:?}:{} fourth message with info n=2 extra=\"extra info\"",
+                    mpath,
+                    start + 4
+                ),
+            ),
+        ];
+        let logs = Log::default();
+        let sub = build(TelioLogLevel::Debug, Box::new(logs.clone()));
```
[the remark this answers is not in the document]
Comment (packgron) [reply src/ffi/logging.rs:214]: done

File: src/ffi/logging.rs (hunk 6 above)
Comment (packgron) [reply src/ffi/logging.rs:66]: done

File: src/ffi/logging.rs (hunk 7 above)
Comment (packgron) [reply src/ffi/logging.rs]: done

File: src/ffi/logging.rs (hunk 10 above)
Comment (packgron) [reply src/ffi/logging.rs]: done

File: src/ffi/logging.rs [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    fn format_event(
+        &self,
+        ctx: &fmt::FmtContext<'_, S, N>,
+        mut writer: fmt::format::Writer<'_>,
+        event: &tracing::Event<'_>,
+    ) -> std::fmt::Result {
+        let meta = event.metadata();
+        write!(
+            writer,
+            "{:?}:{} ",
```
Comment (mathiaspeters) [inline src/ffi/logging.rs:55]: Very minor nit: is `:?` needed here? It looks like it's a string that goes there and strings implement Display. Or am I missing something?

File: src/ffi/logging.rs [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+
+        let actual = logs.0.lock().unwrap().clone();
+
+        assert_eq!(&expected[..], &actual[..])
+    }
+
+    #[derive(Default, Clone, Debug)]
+    struct Log(Arc<Mutex<Vec<(TelioLogLevel, String)>>>);
+    impl TelioLoggerCb for Log {
+        fn log(&self, level: TelioLogLevel, payload: String) -> crate::FFIResult<()> {
```
Comment (stalowyjez) [inline src/ffi/logging.rs:226]: Hmm, it is still `FFI` instead of `Ffi`

File: tests/logger.rs (hunk 2 above)
Comment (stalowyjez) [reply tests/logger.rs:28]: Maybe making some const with that number can be considered, so it will be more apparent?

File: src/ffi/logging.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (stalowyjez) [reply src/ffi/logging.rs:214]: idk, I still see "sub" here ^^'

File: src/ffi/logging.rs [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        // Additional test added to double protect from future breakages
+
+        let msg = from_utf8(buf)
+            // Trim could be avoided by removing writeln! in formatter
+            // But this makes it more univesal for the future, then we dicide to
+            // switch to other formats
+            .map(|msg| msg.trim().to_string())
+            .map_err(|e| io::Error::new(ErrorKind::InvalidData, e))?;
+
+        if let Some(filtered_msg) = filter_log_message(msg) {
```
Comment (stalowyjez) [inline src/ffi/logging.rs:112]: It seems that on `None` here we return buffer length, but log nothing

File: src/ffi/logging.rs (hunk 8 above)
Comment (stalowyjez) [reply src/ffi/logging.rs]: Well, after reading all of this it seems to me that using this `telio-subscriber` crate introduces it more it removes and it it is less straightforward - tbh I prefer the previous implementation.

Comment (packgron) [reply src/ffi/logging.rs]: Previous implementation is fundamentaly broken, unless we implement full span id creation and management

File: src/ffi/logging.rs (hunk 13 above)
Comment (packgron) [reply src/ffi/logging.rs:55]: This debug is not to break how it was displayed previously specificaly "some::path" instead of uquated "some::path"

File: src/ffi/logging.rs (hunk 15 above)
Comment (packgron) [reply src/ffi/logging.rs:112]: Well it would error out looking into write_all that is used on tracing side: https://doc.rust-lang.org/src/std/io/mod.rs.html#1696-1708

File: src/ffi/logging.rs (hunk 14 above)
Comment (packgron) [reply src/ffi/logging.rs:226]: I see lsp did not include #[cfg(test)] and docs also :|

File: tests/logger.rs (hunk 2 above)
Comment (packgron) [reply tests/logger.rs]: added a const

File: src/ffi/logging.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (packgron) [reply src/ffi/logging.rs]: commited

File: src/ffi/logging.rs [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        let actual = logs.0.lock().unwrap().clone();
+
+        assert_eq!(&expected[..], &actual[..])
+    }
+
+    #[derive(Default, Clone, Debug)]
+    struct Log(Arc<Mutex<Vec<(TelioLogLevel, String)>>>);
+    impl TelioLoggerCb for Log {
+        fn log(&self, level: TelioLogLevel, payload: String) -> crate::FfiResult<()> {
+            let mut logs = self.0.lock().expect("ok");
```
Comment (stalowyjez) [inline src/ffi/logging.rs:215]: Nit: I don't feel like "ok" is the right message when sth panics

File: src/ffi/logging.rs [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+use tracing::{level_filters::LevelFilter, Subscriber};
+use tracing_subscriber::{
+    fmt::{self, FormatEvent, FormatFields, MakeWriter},
+    layer::SubscriberExt,
+    registry::LookupSpan,
+};
+
+use crate::{TelioLogLevel, TelioLoggerCb};
+
+pub fn build(log_level: crate::TelioLogLevel, logger: Box<dyn TelioLoggerCb>) -> impl Subscriber {
```
Comment (stalowyjez) [inline src/ffi/logging.rs:16]: I think a public function needs a bit more descriptive name than `build`

File: src/ffi/logging.rs (hunk 16 above)
Comment (packgron) [reply src/ffi/logging.rs]: done

File: src/ffi/logging.rs (hunk 17 above)
Comment (packgron) [reply src/ffi/logging.rs]: done

**Review by stalowyjez — APPROVED** [review]
+1
